### PR TITLE
[Merged by Bors] - feat: port RingTheory.TensorProduct 

### PR DESCRIPTION
**Pull request by kbuzzard**
---
Just an experiment: this is the pushout of the porting PR #3649 with Gabriel's `reenableeta` branch #3414, just to check that we can push it through.

- [x] depends on: #3414

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kbuzzard** [comment]
Bleurgh there's one sorry left but I need to sleep now. If someone else wants to finish the job that would be great!

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#3414
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3414~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/RingTheory/TensorProduct.lean [hunk 1]
```diff
@@ -0,0 +1,1213 @@
...
+@[simp]
+theorem algEquivOfLinearEquivTensorProduct_apply (f w₁ w₂ x) :
+    (algEquivOfLinearEquivTensorProduct f w₁ w₂ : A ⊗[R] B ≃ₐ[R] C) x = f x :=
+  rfl
+#align algebra.tensor_product.alg_equiv_of_linear_equiv_tensor_product_apply Algebra.TensorProduct.algEquivOfLinearEquivTensorProduct_apply
+
+/-- Build an algebra equivalence from a linear equivalence out of a triple tensor product,
+and evidence of multiplicativity on pure tensors.
+-/
+def algEquivOfLinearEquivTripleTensorProduct (f : (A ⊗[R] B) ⊗[R] C ≃ₗ[R] D)
```
Comment (jjaassoonn) [inline Mathlib/RingTheory/TensorProduct.lean:711]: why is this not defined in terms of `algEquivOfLinearEquivTensorProduct`

Comment (jjaassoonn) [reply Mathlib/RingTheory/TensorProduct.lean:711]: Something like this
```lean
example (f : (A ⊗[R] B) ⊗[R] C ≃ₗ[R] D)
    (w₁ :
      ∀ (a₁ a₂ : A) (b₁ b₂ : B) (c₁ c₂ : C),
        f ((a₁ * a₂) ⊗ₜ (b₁ * b₂) ⊗ₜ (c₁ * c₂)) = f (a₁ ⊗ₜ b₁ ⊗ₜ c₁) * f (a₂ ⊗ₜ b₂ ⊗ₜ c₂))
    (w₂ : ∀ r, f (((algebraMap R A) r ⊗ₜ[R] (1 : B)) ⊗ₜ[R] (1 : C)) = (algebraMap R D) r) :
    (A ⊗[R] B) ⊗[R] C ≃ₐ[R] D :=
algEquivOfLinearEquivTensorProduct f (fun x₁ x₂ c₁ c₂ => by
  refine TensorProduct.induction_on x₁ ?_ ?_ ?_ <;>
  refine TensorProduct.induction_on x₂ ?_ ?_ ?_ <;>
  simp only [zero_tmul, tmul_zero, tmul_mul_tmul, map_zero, zero_mul, mul_zero, mul_add, add_mul,
    map_add, add_tmul, tmul_add, w₁] <;>
  try
    intros
    trivial
  . intros ab₁ ab₂ h₁ h₂ a b
    rw [mul_add, add_tmul, map_add, h₁, h₂, map_add, mul_add]
  . intros a b ab₁ ab₂ h₁ h₂
    rw [add_mul, add_tmul, map_add, h₁, h₂, map_add, add_mul]
  . intros ab₁ ab₂ _ _ x y hx hy
    rw [add_mul, add_tmul, map_add, hx, hy, map_add, map_add, mul_add, mul_add, add_mul, mul_add]) <|
  fun r => by
    rw [←w₂]
    congr
```

**Review by ChrisHughes24 — COMMENTED** [review]
One comment, otherwise LGTM

**Comment by semorrison** [comment]
bors merge

File: Mathlib/RingTheory/TensorProduct.lean [hunk 2]
```diff
@@ -574,7 +584,8 @@ variable {A : Type v₁} [Ring A] [Algebra R A]
 variable {B : Type v₂} [Ring B] [Algebra R B]
 
 instance : Ring (A ⊗[R] B) :=
-  { (by infer_instance : AddCommGroup (A ⊗[R] B)), (by infer_instance : Semiring (A ⊗[R] B)) with }
+  { (by infer_instance : Semiring (A ⊗[R] B)) with
+    add_left_neg := add_left_neg }
```
Comment (eric-wieser) [inline Mathlib/RingTheory/TensorProduct.lean]: These lines introduced a diamond in `zsmul` that was fixed in #6141

**Pull request closed without merging**